From bea26e47a592c86df8d0f4c3695220452b4d77ab Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Caol=C3=A1n=20McNamara?= <caolanm@redhat.com>
Date: Wed, 2 Feb 2022 16:21:43 +0000
Subject: [PATCH] rhbz#2047319 drop the SolarMutex during QApplication::init()

https://invent.kde.org/qt/qt/qtwayland/-/merge_requests/24#note_383915

reproducible with with qt5-qtwayland-5.15.2-17.fc35.x86_64 and
export OOO_FORCE_DESKTOP=plasma5
export SAL_USE_VCLPLUGIN=gtk3

Change-Id: Icdf6b8709865c723c266e6400169b33639a1b0e4
---
 shell/Library_kf5be.mk                     |  1 +
 shell/source/backends/kf5be/kf5backend.cxx | 23 ++++++++++++++--------
 2 files changed, 16 insertions(+), 8 deletions(-)

diff --git a/shell/Library_kf5be.mk b/shell/Library_kf5be.mk
index 628145176e0f..157f8b9836f8 100644
--- a/shell/Library_kf5be.mk
+++ b/shell/Library_kf5be.mk
@@ -20,6 +20,7 @@ $(eval $(call gb_Library_use_libraries,kf5be1,\
 	cppu \
 	cppuhelper \
 	sal \
+	vcl \
 ))
 
 $(eval $(call gb_Library_set_componentfile,kf5be1,shell/source/backends/kf5be/kf5be1))
diff --git a/shell/source/backends/kf5be/kf5backend.cxx b/shell/source/backends/kf5be/kf5backend.cxx
index e448c124d88a..bde5f85c1ae4 100644
--- a/shell/source/backends/kf5be/kf5backend.cxx
+++ b/shell/source/backends/kf5be/kf5backend.cxx
@@ -42,6 +42,7 @@
 #include <rtl/ustring.hxx>
 #include <sal/types.h>
 #include <uno/current_context.hxx>
+#include <vcl/svapp.hxx>
 
 #include <osl/process.h>
 #include <osl/thread.h>
@@ -170,14 +171,20 @@ void initQApp(std::map<OUString, css::beans::Optional<css::uno::Any>>& rSettings
         unsetenv("SESSION_MANAGER");
     }
 
-    std::unique_ptr<QApplication> app(new QApplication(nFakeArgc, pFakeArgv));
-    QObject::connect(app.get(), &QObject::destroyed, app.get(), [nFakeArgc, pFakeArgv]() {
-        for (int i = 0; i < nFakeArgc; ++i)
-            free(pFakeArgv[i]);
-        delete[] pFakeArgv;
-    });
-
-    readKDESettings(rSettings);
+    {
+        // rhbz#2047319 drop the SolarMutex during the execution of QApplication::init()
+        // https://invent.kde.org/qt/qt/qtwayland/-/merge_requests/24#note_383915
+        SolarMutexReleaser aReleaser; // rhbz#2047319 drop the SolarMutex during the execution
+
+        std::unique_ptr<QApplication> app(new QApplication(nFakeArgc, pFakeArgv));
+        QObject::connect(app.get(), &QObject::destroyed, app.get(), [nFakeArgc, pFakeArgv]() {
+            for (int i = 0; i < nFakeArgc; ++i)
+                free(pFakeArgv[i]);
+            delete[] pFakeArgv;
+        });
+
+        readKDESettings(rSettings);
+    }
 
     if (session_manager != nullptr)
     {
-- 
2.33.1

